### ISSUE TITLE
Unigram mixture

### ISSUE THREAD

**Issue by jcapde87** [blei-lab/edward#361]
@bertini36 and me were wondering if we could develop topic models in Edward, but we couldn't find any example in the repository yet. Do you foresee any stopper to have these models in Edward through score gradients? 

We've started playing around with a simple unigram mixture model and learn some things that we'd like to share with you. In case you think that these findings are relevant, we'll be happy to submit a PR with the 'tf_unigram_mixture.py` example and keep investigating the issues.

- Regarding how to represent documents, we've initially thought to use a matrix representation (N(docs)xW(words)), but we've discarded because it'd be too sparse. We've then opted to represent documents in a vectorial format, in which each element indexes a word in the vocabulary. Documents, with different length, are appended one after the other in this vector. To keep track of which words belong to each document, we had to create an `indexes` vector that contained the document index for each word.

```
    #data generation
    w = []                                     
    indexes = []
    for n in range(N):
        c[n] = np.random.choice(K, 1, p=pi)[0]
        for m in xrange(Mn[n]):
            aux_word = np.random.choice(M, 1, p=betas[c[n], :])[0]
            w.append(aux_word)
            indexes.append(n)
```

-  Because of this data representation, we couldn't use batches since the current procedure slices data into mini batches with exact size and hence some documents could fall in different batches. We believe that we should look for another procedure to create batches using the `indexes` vector (maybe implement a  `next_batch()` function such as in `probabilistic_pca_subsampling.py`).   
 
-  We coded the `log_prob` function and used the `tf.gather()` and `tf.segment_sum()` to index and sum over the words in each document in the `log_lik` term
```
    matrix = []
    for k in range(self.K):
        logb_kwnm = tf.log(tf.gather(tf.transpose(betas[k]), tf.to_int32(w)))  
        n_tensor = tf.segment_sum(logb_kwnm, tf.to_int32(indexes))     
        matrix += [tf.ones(self.N) * tf.log(pi[k]) + n_tensor]
    matrix = tf.pack(matrix)
    vector = log_sum_exp(matrix, 0)
    log_lik = tf.reduce_sum(vector)
```

- The `log_prior` term resembles that of `tf_mixture_gaussian.py`, except for the fact that here we had to iterate over each mixture component to compute the logpdf.

```
    log_prior = dirichlet.logpdf(pi, self.alpha)
    for k in xrange(self.K):
      log_prior += dirichlet.logpdf(betas[k], self.alpha_beta)
```

- With the model above, we've got `InvalidArgumentError` because there were Inf values in the expression:
```
vector = log_sum_exp(matrix, 0)
```

- The cause of the above infinities was some 0's in the sampled `betas`, which became Inf after the tf.log(). The sampled  `betas` had 0's because the Dirichlets had a wrong parameterization since some of their parameters were also zero. We believe that this was due to the fact that `tf.nn.softplus()` failed to represent some large negative values because of the lack of float precision. A quick and dirty way to avoid the infinites, was to simply add 0.1 to the output of operation `tf.nn.softplus()`. 

```
tf.add(tf.nn.softplus(tf.Variable(tf.random_normal([M],dtype=tf.float64))), 0.1) 
```

- With the above changes, the unigram mixture for two components and for some synthetic data  `betas` converges and recovers the mixture components: 

```
    betas = np.array([[0.2, 0.2, 0.2, 0.2, 0.2, 0.0, 0.0, 0.0, 0.0, 0.0],
                      [0.0, 0.0, 0.0, 0.0, 0.0, 0.2, 0.2, 0.2, 0.2, 0.2]])
```
![inferred](https://cloud.githubusercontent.com/assets/4015373/20644555/799eb5e4-b438-11e6-9f71-9fec3a6c1918.png)
